### PR TITLE
[OCPCLOUD-1423] Machine webhook validation for Azure Ultra Disks

### PR DESCRIPTION
**Pull request by damdo**
- it bumps the openshift/api to get the latest machine/v1beta1 api changes introduced by https://github.com/openshift/api/pull/1119 
- it adds machine webhook validation for the Azure Ultra Disks feature
- it adds tests for machine webhook validation for the Azure Ultra Disks feature

### CODE REVIEW

**Comment by damdo** [comment]
Note: I did some manual testing of the webhook validation:
```
oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].lun: Invalid value: 0: each Data Disk must have a unique lun): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].lun: Invalid value: 0: each Data Disk must have a unique lun

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].nameSuffix: Invalid value: "testdiskk": each Data Disk must have a unique nameSuffix): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].nameSuffix: Invalid value: "testdiskk": each Data Disk must have a unique nameSuffix

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].nameSuffix: Invalid value: "testdiskkdaoudasuigdsaiugdsaiusdgisaugdiuasgdsaiudgasiudgasiudgasiudgsaiudgasdisaiduashdiuasgdsaiugdasuidgiuasgdi": too long, the overall disk name must not exceed 80 chars): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].nameSuffix: Invalid value: "testdiskkdaoudasuigdsaiugdsaiusdgisaugdiuasgdsaiudgasiudgasiudgasiudgsaiudgasdisaiduashdiuasgdsaiugdasuidgiuasgdi": too long, the overall disk name must not exceed 80 chars

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss$aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss$aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens
oc -n openshift-machine-api apply -f ms-n1.yaml

Error from server (providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss$aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss$aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss^aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].nameSuffix: Invalid value: "sss^aaa": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].diskSizeGB: Invalid value: 1: diskSizeGB must be provided and at least 4GB in size): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].diskSizeGB: Invalid value: 1: diskSizeGB must be provided and at least 4GB in size

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].cachingType: Invalid value: "ReadOnly": must be "None" or omitted when storageAccountType is "UltraSSD_LRS"): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].cachingType: Invalid value: "ReadOnly": must be "None" or omitted when storageAccountType is "UltraSSD_LRS"

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].lun: Invalid value: 64: must be higher or equal than 0 and lower than 64): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].lun: Invalid value: 64: must be higher or equal than 0 and lower than 64

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[1].lun: Invalid value: -1: must be higher or equal than 0 and lower than 64): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[1].lun: Invalid value: -1: must be higher or equal than 0 and lower than 64

oc -n openshift-machine-api apply -f ms-n1.yaml
Error from server (providerSpec.dataDisks[0].nameSuffix: Invalid value: "": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens): error when creating "ms-n1.yaml": admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.dataDisks[0].nameSuffix: Invalid value: "": nameSuffix must be provided and can only contain letters, numbers, underscores, periods or hyphens
```

**Review by alexander-demichev — APPROVED** [review]
/approve

File: pkg/webhooks/machine_webhook.go [hunk 1]
```diff
@@ -1332,6 +1335,58 @@ func validateMachineLifecycleHooks(m, oldM *machinev1.Machine) []error {
 	return errs
 }
 
+func validateAzureDataDisks(machineName string, spec *machinev1.AzureMachineProviderSpec, parentPath *field.Path) []error {
+
+	var errs []error
+	dataDiskLuns := make(map[int32]struct{})
+	dataDiskNames := make(map[string]struct{})
+	reg := regexp.MustCompile(`^[a-zA-Z0-9](?:[\w\.-]*[a-zA-Z0-9])?$`)
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1343]: Please add a comment to explain what the regex is trying to achieve, reading regexs and interpreting them takes a lot longer than reading a well thought out explanation of the regex

File: pkg/webhooks/machine_webhook.go [hunk 2]
```diff
@@ -1332,6 +1335,58 @@ func validateMachineLifecycleHooks(m, oldM *machinev1.Machine) []error {
 	return errs
 }
 
+func validateAzureDataDisks(machineName string, spec *machinev1.AzureMachineProviderSpec, parentPath *field.Path) []error {
+
+	var errs []error
+	dataDiskLuns := make(map[int32]struct{})
+	dataDiskNames := make(map[string]struct{})
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1342]: Good use of empty struct for an efficient set!

File: pkg/webhooks/machine_webhook.go [hunk 3]
```diff
@@ -1332,6 +1335,58 @@ func validateMachineLifecycleHooks(m, oldM *machinev1.Machine) []error {
...
+		if _, exists := dataDiskNames[disk.NameSuffix]; exists {
+			errs = append(errs, field.Invalid(fldPath.Child("nameSuffix"), disk.NameSuffix, "each Data Disk must have a unique nameSuffix"))
+		}
+
+		if disk.DiskSizeGB < 4 {
+			errs = append(errs, field.Invalid(fldPath.Child("diskSizeGB"), disk.DiskSizeGB, "diskSizeGB must be provided and at least 4GB in size"))
+		}
+
+		if disk.Lun < 0 || disk.Lun > 63 {
+			errs = append(errs, field.Invalid(fldPath.Child("lun"), disk.Lun, "must be higher or equal than 0 and lower than 64"))
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1367]: Nit: Typically we would say `greater than or equal to 0`
```suggestion
			errs = append(errs, field.Invalid(fldPath.Child("lun"), disk.Lun, "must be greater than or equal to 0 and less than 64"))
```

File: pkg/webhooks/machine_webhook.go (hunk 2 above)
Comment (damdo) [reply pkg/webhooks/machine_webhook.go:1342]: 🙂

File: pkg/webhooks/machine_webhook.go (hunk 1 above)
Comment (damdo) [reply pkg/webhooks/machine_webhook.go:1345]: Yes I have a comment explaining what it does on the api spec, but forgot to add it here. Thanks for flagging.
Added.

File: pkg/webhooks/machine_webhook.go (hunk 3 above)
Comment (damdo) [reply pkg/webhooks/machine_webhook.go]: Yes, it definitely sounds better 👍 Updated.

**Comment by damdo** [comment]
Hold until we confirm the desired behaviour on instance deletion. 
/hold

**Pull request merged**